File: packages/search-client-python/src/aacsearch/_admin_client.py
"""
Server-side administration client for the AACSearch v1 REST API.

Requires an admin-scope API key (``aa_admin_*``). Never bundle in browser code.
"""

from __future__ import annotations

from typing import Any

from ._transport import _request


class AdminClient:
    """Client for managing AACSearch projects, indexes, documents, keys, and more.

    Args:
        base_url: Origin of the AACSearch deployment (e.g. ``https://app.aacsearch.com``).
        api_key: An admin-scope API key (``aa_admin_…``).
        project_id: The project (organization) ID. Discover via ``get_project()`` if unknown.
        timeout: HTTP request timeout in seconds.
    """

    def __init__(
        self,
        base_url: str,
        api_key: str,
        project_id: str,
        timeout: float = 30.0,
    ) -> None:
        if not base_url:
            raise ValueError("AdminClient: base_url is required")
        if not api_key:
            raise ValueError("AdminClient: api_key is required")
        if not project_id:
            raise ValueError("AdminClient: project_id is required")

        self._base_url = base_url
        self._api_key = api_key
        self._project_id = project_id
        self._timeout = timeout

    # ── Project operations ───────────────────────────────────────────

    def get_project(self) -> dict[str, Any]:
        """Get current project details."""
        return _request(self._base_url, self._api_key, "GET", "/api/v1/projects", timeout=self._timeout)

    def create_project(self, name: str, slug: str, logo: str | None = None) -> dict[str, Any]:
        """Create a new project (organization).

        Args:
            name: Display name (max 120 chars).
            slug: URL-safe identifier (``^[a-z0-9][a-z0-9-]*$``, max 64 chars).
            logo: Optional logo URL.
        """
        body: dict[str, Any] = {"name": name, "slug": slug}
        if logo is not None:
            body["logo"] = logo
        return _request(self._base_url, self._api_key, "POST", "/api/v1/projects", body=body, timeout=self._timeout)

    def get_project_by_id(self, project_id: str) -> dict[str, Any]:
        """Get project by ID.

        Args:
            project_id: The project (organization) ID.
        """
        return _request(self._base_url, self._api_key, "GET", f"/api/v1/projects/{_e(project_id)}", timeout=self._timeout)

    # ── Index operations ─────────────────────────────────────────────

    def list_indexes(self) -> list[dict[str, Any]]:
        """List all indexes in the project."""
        return _request(self._base_url, self._api_key, "GET", f"/api/v1/projects/{_e(self._project_id)}/indexes", timeout=self._timeout)

    def get_index(self, index_id: str) -> dict[str, Any]:
        """Get a single index by its ID."""
        return _request(self._base_url, self._api_key, "GET", f"/api/v1/indexes/{_e(index_id)}", timeout=self._timeout)

    def create_index(
        self,
        slug: str,
        display_name: str,
        fields: list[dict[str, Any]],
        default_sorting_field: str | None = None,
    ) -> dict[str, Any]:
        """Create a new search index with schema and Typesense collection.

        Args:
            slug: URL-safe index identifier.
            display_name: Human-readable name.
            fields: Array of field definitions (``name``, ``type``, ``facet``, ``optional``, etc.).
            default_sorting_field: Optional default sort field name.
        """
        body: dict[str, Any] = {"slug": slug, "displayName": display_name, "fields": fields}
        if default_sorting_field is not None:
            body["defaultSortingField"] = default_sorting_field
        return _request(self._base_url, self._api_key, "POST", f"/api/v1/projects/{_e(self._project_id)}/indexes", body=body, timeout=self._timeout)

    def update_index(self, index_id: str, *, display_name: str | None = None, enabled: bool | None = None) -> dict[str, Any]:
        """Update index display name and/or enabled status."""
        body: dict[str, Any] = {}
        if display_name is not None:
            body["displayName"] = display_name
        if enabled is not None:
            body["enabled"] = enabled
        return _request(self._base_url, self._api_key, "PATCH", f"/api/v1/indexes/{_e(index_id)}", body=body, timeout=self._timeout)

    def delete_index(self, index_id: str) -> dict[str, Any]:
        """Delete an index and its Typesense collections."""
        return _request(self._base_url, self._api_key, "DELETE", f"/api/v1/indexes/{_e(index_id)}", timeout=self._timeout)

    def get_index_stats(self, index_id: str) -> dict[str, Any]:
        """Get index statistics (document count, usage, ingest queue)."""
        return _request(self._base_url, self._api_key, "GET", f"/api/v1/indexes/{_e(index_id)}/stats", timeout=self._timeout)

    # ── Document operations ──────────────────────────────────────────

    def list_documents(
        self,
        index_id: str,
        *,
        q: str | None = None,
        page: int | None = None,
        per_page: int | None = None,
        filter_by: str | None = None,
    ) -> dict[str, Any]:
        """Browse / list documents in an index.

        Args:
            index_id: The index ID.
            q: Query string (default ``"*"``).
            page: Page number (1-based).
            per_page: Results per page (max 250).
            filter_by: Filter expression.
        """
        params: dict[str, str] = {}
        if q is not None:
            params["q"] = q
        if page is not None:
            params["page"] = str(page)
        if per_page is not None:
            params["perPage"] = str(per_page)
        if filter_by is not None:
            params["filterBy"] = filter_by
        return _request(
            self._base_url, self._api_key, "GET", f"/api/v1/indexes/{_e(index_id)}/documents",
            params=params or None, timeout=self._timeout,
        )

    def upsert_document(self, index_id: str, document_id: str, document: dict[str, Any]) -> dict[str, Any]:
        """Upsert a single document."""
        return _request(
            self._base_url, self._api_key, "PUT",
            f"/api/v1/indexes/{_e(index_id)}/documents/{_e(document_id)}",
            body=document, timeout=self._timeout,
        )

    def batch_upsert_documents(self, index_id: str, documents: list[dict[str, Any]]) -> dict[str, Any]:
        """Batch upsert documents (up to 5000 at a time)."""
        return _request(
            self._base_url, self._api_key, "POST",
            f"/api/v1/indexes/{_e(index_id)}/documents:batch",
            body={"documents": documents}, timeout=self._timeout,
        )

    def batch_delete_documents(self, index_id: str, ids: list[str]) -> dict[str, Any]:
        """Batch delete documents by IDs."""
        return _request(
            self._base_url, self._api_key, "POST",
            f"/api/v1/indexes/{_e(index_id)}/documents:batchDelete",
            body={"ids": ids}, timeout=self._timeout,
        )

    def delete_document(self, index_id: str, document_id: str) -> dict[str, Any]:
        """Delete a single document."""
        return _request(
            self._base_url, self._api_key, "DELETE",
            f"/api/v1/indexes/{_e(index_id)}/documents/{_e(document_id)}",
            timeout=self._timeout,
        )

    # ── Search ───────────────────────────────────────────────────────

    def search(self, index_id: str, *, q: str = "*", **kwargs: Any) -> dict[str, Any]:
        """Search an index (requires a key with search scope).

        Args:
            index_id: The index ID or slug.
            q: Search query (default ``"*"``).
            **kwargs: Additional search parameters (query_by, filter_by, facet_by,
                sort_by, per_page, page, highlight_fields, etc.).
        """
        params: dict[str, Any] = {"q": q}
        params.update(kwargs)
        return _request(
            self._base_url, self._api_key, "POST",
            f"/api/v1/indexes/{_e(index_id)}/search",
            body=params, timeout=self._timeout,
        )

    # ── API Key operations ───────────────────────────────────────────

    def list_keys(self) -> list[dict[str, Any]]:
        """List all API keys in the project."""
        return _request(
            self._base_url, self._api_key, "GET",
            f"/api/v1/projects/{_e(self._project_id)}/keys",
            timeout=self._timeout,
        )

    def create_key(
        self,
        name: str,
        scopes: list[str],
        index_slug: str,
        *,
        allowed_origins: list[str] | None = None,
        rate_limit_per_minute: int | None = None,
    ) -> dict[str, Any]:
        """Create a new API key.

        Args:
            name: Human-readable key name.
            scopes: Permission scopes (e.g. ``["search"]``, ``["ingest"]``, ``["admin"]``).
            index_slug: The index slug the key should be scoped to.
            allowed_origins: Optional CORS origins list.
            rate_limit_per_minute: Optional rate limit override.
        """
        body: dict[str, Any] = {"name": name, "scopes": scopes, "indexSlug": index_slug}
        if allowed_origins is not None:
            body["allowedOrigins"] = allowed_origins
        if rate_limit_per_minute is not None:
            body["rateLimitPerMinute"] = rate_limit_per_minute
        return _request(
            self._base_url, self._api_key, "POST",
            f"/api/v1/projects/{_e(self._project_id)}/keys",
            body=body, timeout=self._timeout,
        )

    def revoke_key(self, key_id: str) -> dict[str, Any]:
        """Revoke an API key by its ID."""
        return _request(
            self._base_url, self._api_key, "DELETE",
            f"/api/v1/keys/{_e(key_id)}",
            timeout=self._timeout,
        )

    # ── Analytics ────────────────────────────────────────────────────

    def get_analytics(self, period: str | None = None) -> dict[str, Any]:
        """Get aggregated search analytics for the project.

        Args:
            period: Time period (e.g. ``"24h"``, ``"7d"``, ``"30d"``).
        """
        if period:
            return _request(
                self._base_url, self._api_key, "GET",
                f"/api/v1/projects/{_e(self._project_id)}/analytics",
                params={"period": period}, timeout=self._timeout,
            )
        return _request(
            self._base_url, self._api_key, "GET",
            f"/api/v1/projects/{_e(self._project_id)}/analytics",
            timeout=self._timeout,
        )

    def get_usage(self, window_days: int | None = None) -> dict[str, Any]:
        """Get raw usage data for the project.

        Args:
            window_days: Number of days of usage data to return.
        """
        if window_days is not None:
            return _request(
                self._base_url, self._api_key, "GET",
                f"/api/v1/projects/{_e(self._project_id)}/usage",
                params={"windowDays": str(window_days)}, timeout=self._timeout,
            )
        return _request(
            self._base_url, self._api_key, "GET",
            f"/api/v1/projects/{_e(self._project_id)}/usage",
            timeout=self._timeout,
        )

    # ── Synonym operations ───────────────────────────────────────────

    def list_synonyms(self, index_id: str) -> list[dict[str, Any]]:
        """List synonyms for an index."""
        return _request(
            self._base_url, self._api_key, "GET",
            f"/api/v1/indexes/{_e(index_id)}/synonyms",
            timeout=self._timeout,
        )

    def upsert_synonyms(self, index_id: str, synonyms: list[dict[str, str]]) -> dict[str, Any]:
        """Replace all synonyms for an index.

        Args:
            index_id: The index ID.
            synonyms: List of ``{"root": ..., "synonym": ...}`` pairs.
        """
        return _request(
            self._base_url, self._api_key, "PUT",
            f"/api/v1/indexes/{_e(index_id)}/synonyms",
            body={"synonyms": synonyms}, timeout=self._timeout,
        )

    # ── Curation operations ──────────────────────────────────────────

    def list_curations(self, index_id: str) -> list[dict[str, Any]]:
        """List curations (overrides) for an index."""
        return _request(
            self._base_url, self._api_key, "GET",
            f"/api/v1/indexes/{_e(index_id)}/curations",
            timeout=self._timeout,
        )

    def upsert_curations(self, index_id: str, curations: list[dict[str, Any]]) -> dict[str, Any]:
        """Replace all curations for an index.

        Args:
            index_id: The index ID.
            curations: List of curation rules with ``query``, ``pinnedIds``, ``hiddenIds``.
        """
        return _request(
            self._base_url, self._api_key, "PUT",
            f"/api/v1/indexes/{_e(index_id)}/curations",
            body={"curations": curations}, timeout=self._timeout,
        )


def _e(value: str) -> str:
    """URL-encode a path segment."""
    from urllib.parse import quote
    return quote(value, safe="")

File: packages/search-client-ruby/lib/aacsearch/admin_client.rb
# frozen_string_literal: true

require_relative "client"

module Aacsearch
  # Server-side administration client for the AACsearch v2 REST API.
  #
  # Requires an admin-scope API key (aa_admin_*). Never bundle in browser
  # or mobile application code.
  #
  # Provides full CRUD for projects, indexes, documents, API keys, synonyms,
  # curations, and read-only access to analytics and usage data.
  #
  # @example
  #   admin = Aacsearch::AdminClient.new(
  #     base_url: "https://api.aacsearch.com",
  #     api_key: "aa_admin_abc123"
  #   )
  #
  #   # Get current project
  #   project = admin.get_project
  #
  #   # List indexes in a project
  #   indexes = admin.list_indexes(project["id"])
  class AdminClient
    # Create a new v2 admin client.
    #
    # @param base_url [String] Origin of the AACsearch deployment
    #   (e.g. "https://api.aacsearch.com").
    # @param api_key [String] An admin-scope API key (aa_admin_*).
    # @param timeout [Integer] HTTP request timeout in seconds (default: 30).
    def initialize(base_url:, api_key:, timeout: Client::DEFAULT_TIMEOUT)
      @client = Client.new(base_url: base_url, api_key: api_key, timeout: timeout)
    end

    # ── Project operations ──────────────────────────────────────────────────

    # Get the current project (organization) details.
    #
    # GET /api/v2/projects
    #
    # @return [Hash] Project details (see {ProjectFields}).
    def get_project
      @client.request(:get, "/api/v2/projects")
    end

    # Create a new project (organization).
    #
    # POST /api/v2/projects
    #
    # @param name [String] Display name (max 120 chars).
    # @param slug [String] URL-safe identifier (lowercase, digits, hyphens).
    # @param logo [String, nil] Optional URL to project logo image.
    # @return [Hash] Created project details.
    def create_project(name, slug, logo: nil)
      body = { "name" => name, "slug" => slug }
      body["logo"] = logo if logo
      @client.request(:post, "/api/v2/projects", body)
    end

    # Get a project by its ID.
    #
    # GET /api/v2/projects/{projectId}
    #
    # @param project_id [String] The project ID.
    # @return [Hash] Project details.
    def get_project_by_id(project_id)
      @client.request(:get, "/api/v2/projects/#{url_encode(project_id)}")
    end

    # ── Index operations ────────────────────────────────────────────────────

    # List all search indexes in a project.
    #
    # GET /api/v2/projects/{projectId}/indexes
    #
    # @param project_id [String] The project ID.
    # @return [Array<Hash>] List of search indexes.
    def list_indexes(project_id)
      @client.request(:get, "/api/v2/projects/#{url_encode(project_id)}/indexes")
    end

    # Create a new search index.
    #
    # POST /api/v2/projects/{projectId}/indexes
    #
    # @param project_id [String] The project ID.
    # @param slug [String] URL-safe index identifier.
    # @param display_name [String] Human-readable name (max 120 chars).
    # @param fields [Array<Hash>] Array of field definitions. Each field
    #   requires +name+ and +type+; may include +facet+, +index+, +optional+,
    #   +locale+, +sort+, +infix+, +numDimensions+.
    # @param default_sorting_field [String, nil] Optional default sort field.
    # @return [Hash] Created search index details.
    def create_index(project_id, slug, display_name, fields, default_sorting_field: nil)
      body = { "slug" => slug, "displayName" => display_name, "fields" => fields }
      body["defaultSortingField"] = default_sorting_field if default_sorting_field
      @client.request(:post, "/api/v2/projects/#{url_encode(project_id)}/indexes", body)
    end

    # Get a single search index by its ID.
    #
    # GET /api/v2/indexes/{indexId}
    #
    # @param index_id [String] The index ID.
    # @return [Hash] Index details.
    def get_index(index_id)
      @client.request(:get, "/api/v2/indexes/#{url_encode(index_id)}")
    end

    # Update index metadata.
    #
    # PATCH /api/v2/indexes/{indexId}
    #
    # @param index_id [String] The index ID.
    # @param display_name [String, nil] New display name.
    # @param enabled [Boolean, nil] Whether the index is enabled.
    # @param default_sorting_field [String, nil] New default sort field.
    # @return [Hash] Updated index details.
    def update_index(index_id, display_name: nil, enabled: nil, default_sorting_field: nil)
      body = {}
      body["displayName"] = display_name unless display_name.nil?
      body["enabled"] = enabled unless enabled.nil?
      body["defaultSortingField"] = default_sorting_field unless default_sorting_field.nil?
      @client.request(:patch, "/api/v2/indexes/#{url_encode(index_id)}", body)
    end

    # Permanently delete a search index and its Typesense collection.
    #
    # DELETE /api/v2/indexes/{indexId}
    #
    # @param index_id [String] The index ID.
    # @return [Hash, nil] Response body (may be nil for 204).
    def delete_index(index_id)
      @client.request(:delete, "/api/v2/indexes/#{url_encode(index_id)}")
    end

    # Get index statistics (document count, usage, ingest queue).
    #
    # GET /api/v2/indexes/{indexId}/stats
    #
    # @param index_id [String] The index ID.
    # @return [Hash] Index statistics.
    def get_index_stats(index_id)
      @client.request(:get, "/api/v2/indexes/#{url_encode(index_id)}/stats")
    end

    # ── Document operations (v2, cursor-based pagination) ───────────────────

    # List documents in an index with cursor-based pagination.
    #
    # GET /api/v2/indexes/{indexId}/documents
    #
    # @param index_id [String] The index ID.
    # @param cursor [String, nil] Pagination cursor from previous response.
    # @param per_page [Integer, nil] Results per page (max 100, default 20).
    # @param filter_by [String, nil] Typesense filter expression.
    # @return [Hash] Paginated document list with +hits+, +found+, +cursor+.
    def list_documents(index_id, cursor: nil, per_page: nil, filter_by: nil)
      params = {}
      params["cursor"] = cursor if cursor
      params["perPage"] = per_page.to_s if per_page
      params["filterBy"] = filter_by if filter_by
      @client.request(:get, "/api/v2/indexes/#{url_encode(index_id)}/documents", nil, params)
    end

    # Upsert documents into an index (batch, async via ingest buffer).
    #
    # POST /api/v2/indexes/{indexId}/documents
    #
    # @param index_id [String] The index ID.
    # @param documents [Array<Hash>] Array of document objects. Each should
    #   have a unique +id+ field.
    # @param action [String] Import action: +"upsert"+ (default), +"create"+,
    #   or +"update"+.
    # @return [Hash] Response with queued/accepted counts.
    def upsert_documents(index_id, documents, action: "upsert")
      body = { "documents" => documents, "action" => action }
      @client.request(:post, "/api/v2/indexes/#{url_encode(index_id)}/documents", body)
    end

    # Get a single document by its ID.
    #
    # GET /api/v2/indexes/{indexId}/documents/{documentId}
    #
    # @param index_id [String] The index ID.
    # @param document_id [String] The document ID.
    # @return [Hash] Document content.
    def get_document(index_id, document_id)
      @client.request(:get, "/api/v2/indexes/#{url_encode(index_id)}/documents/#{url_encode(document_id)}")
    end

    # Upsert a single document.
    #
    # PUT /api/v2/indexes/{indexId}/documents/{documentId}
    #
    # @param index_id [String] The index ID.
    # @param document_id [String] The document ID.
    # @param data [Hash] Document data (does not need to include +id+).
    # @return [Hash] Upserted document.
    def upsert_document(index_id, document_id, data)
      @client.request(:put, "/api/v2/indexes/#{url_encode(index_id)}/documents/#{url_encode(document_id)}", data)
    end

    # Delete a single document by its ID.
    #
    # DELETE /api/v2/indexes/{indexId}/documents/{documentId}
    #
    # @param index_id [String] The index ID.
    # @param document_id [String] The document ID.
    # @return [Hash, nil] Response body.
    def delete_document(index_id, document_id)
      @client.request(:delete, "/api/v2/indexes/#{url_encode(index_id)}/documents/#{url_encode(document_id)}")
    end

    # Batch delete documents by their IDs.
    #
    # POST /api/v2/indexes/{indexId}/documents:batchDelete
    #
    # @param index_id [String] The index ID.
    # @param ids [Array<String>] Array of document IDs to delete.
    # @return [Hash] Deletion result.
    def batch_delete_documents(index_id, ids)
      body = { "ids" => ids }
      @client.request(:post, "/api/v2/indexes/#{url_encode(index_id)}/documents:batchDelete", body)
    end

    # Export documents from an index.
    #
    # GET /api/v2/indexes/{indexId}/documents:export
    #
    # @param index_id [String] The index ID.
    # @param format [String] Export format: +"jsonl"+ (default) or +"json"+.
    # @param filter_by [String, nil] Optional Typesense filter expression.
    # @return [Array, String] Exported documents.
    def export_documents(index_id, format: "jsonl", filter_by: nil)
      params = { "format" => format }
      params["filterBy"] = filter_by if filter_by
      @client.request(:get, "/api/v2/indexes/#{url_encode(index_id)}/documents:export", nil, params)
    end

    # ── Search (admin, requires search-scoped key) ──────────────────────────

    # Search an index (requires a key with search scope).
    #
    # POST /api/v2/indexes/{indexId}/search
    #
    # @param index_id [String] The index ID or slug.
    # @param params [Hash] Search parameters (same as {SearchClient#search}).
    # @return [Hash] Search results.
    def search(index_id, params = {})
      body = params.empty? ? { "q" => "*" } : params
      @client.request(:post, "/api/v2/indexes/#{url_encode(index_id)}/search", body)
    end

    # Execute multiple search queries in a single request.
    #
    # POST /api/v2/multi-search
    #
    # @param searches [Array<Hash>] List of search definitions.
    # @return [Hash] Combined search results.
    def multi_search(searches)
      @client.request(:post, "/api/v2/multi-search", searches)
    end

    # ── API Key operations ──────────────────────────────────────────────────

    # List all API keys in the project.
    #
    # GET /api/v2/projects/{projectId}/keys
    #
    # @param project_id [String] The project ID.
    # @return [Array<Hash>] List of API keys.
    def list_keys(project_id)
      @client.request(:get, "/api/v2/projects/#{url_encode(project_id)}/keys")
    end

    # Create a new API key for the project.
    #
    # POST /api/v2/projects/{projectId}/keys
    #
    # @param project_id [String] The project ID.
    # @param description [String] Human-readable key description.
    # @param scopes [Array<String>] Permission scopes (e.g. +["search"]+,
    #   +["ingest"]+, +["admin"]+).
    # @param rate_limit_per_minute [Integer, nil] Optional rate limit override.
    # @param allowed_origins [Array<String>, nil] Optional CORS origins.
    # @return [Hash] Created API key (includes +rawKey+ shown once).
    def create_key(project_id, description, scopes, rate_limit_per_minute: nil, allowed_origins: nil)
      body = { "description" => description, "scopes" => scopes }
      body["rateLimitPerMinute"] = rate_limit_per_minute if rate_limit_per_minute
      body["allowedOrigins"] = allowed_origins if allowed_origins
      @client.request(:post, "/api/v2/projects/#{url_encode(project_id)}/keys", body)
    end

    # Revoke an API key by its ID.
    #
    # DELETE /api/v2/keys/{keyId}
    #
    # @param key_id [String] The API key ID.
    # @return [Hash, nil] Revocation result.
    def revoke_key(key_id)
      @client.request(:delete, "/api/v2/keys/#{url_encode(key_id)}")
    end

    # ── Synonym operations (v2) ─────────────────────────────────────────────

    # List all synonyms for an index.
    #
    # GET /api/v2/indexes/{indexId}/synonyms
    #
    # @param index_id [String] The index ID.
    # @return [Array<Hash>] List of synonyms.
    def list_synonyms(index_id)
      @client.request(:get, "/api/v2/indexes/#{url_encode(index_id)}/synonyms")
    end

    # Create a synonym for an index.
    #
    # POST /api/v2/indexes/{indexId}/synonyms
    #
    # @param index_id [String] The index ID.
    # @param synonym [Hash] Synonym definition with +root+, +synonyms+, +locale+.
    # @return [Hash] Created synonym.
    def create_synonym(index_id, synonym)
      @client.request(:post, "/api/v2/indexes/#{url_encode(index_id)}/synonyms", synonym)
    end

    # Upsert (batch replace) all synonyms for an index.
    #
    # PUT /api/v2/indexes/{indexId}/synonyms
    #
    # @param index_id [String] The index ID.
    # @param synonyms [Array<Hash>] Array of synonym definitions.
    # @return [Array<Hash>] Upserted synonyms.
    def upsert_synonyms(index_id, synonyms)
      @client.request(:put, "/api/v2/indexes/#{url_encode(index_id)}/synonyms", synonyms)
    end

    # Delete a synonym by its ID.
    #
    # DELETE /api/v2/indexes/{indexId}/synonyms?synonymId={synonymId}
    #
    # @param index_id [String] The index ID.
    # @param synonym_id [String] The synonym ID to delete.
    # @return [Hash, nil] Deletion result.
    def delete_synonym(index_id, synonym_id)
      params = { "synonymId" => synonym_id }
      @client.request(:delete, "/api/v2/indexes/#{url_encode(index_id)}/synonyms", nil, params)
    end

    # ── Curation operations (v2) ────────────────────────────────────────────

    # List all curations (search overrides) for an index.
    #
    # GET /api/v2/indexes/{indexId}/curations
    #
    # @param index_id [String] The index ID.
    # @return [Array<Hash>] List of curations.
    def list_curations(index_id)
      @client.request(:get, "/api/v2/indexes/#{url_encode(index_id)}/curations")
    end

    # Create a curation (search override) for an index.
    #
    # POST /api/v2/indexes/{indexId}/curations
    #
    # @param index_id [String] The index ID.
    # @param curation [Hash] Curation definition with +query+, +pinnedIds+,
    #   +hiddenIds+, +boostedIds+.
    # @return [Hash] Created curation.
    def create_curation(index_id, curation)
      @client.request(:post, "/api/v2/indexes/#{url_encode(index_id)}/curations", curation)
    end

    # Upsert (batch replace) all curations for an index.
    #
    # PUT /api/v2/indexes/{indexId}/curations
    #
    # @param index_id [String] The index ID.
    # @param curations [Array<Hash>] Array of curation definitions.
    # @return [Array<Hash>] Upserted curations.
    def upsert_curations(index_id, curations)
      @client.request(:put, "/api/v2/indexes/#{url_encode(index_id)}/curations", curations)
    end

    # ── Analytics ───────────────────────────────────────────────────────────

    # Get aggregated search analytics for the project.
    #
    # GET /api/v2/projects/{projectId}/analytics
    #
    # @param project_id [String] The project ID.
    # @return [Hash] Analytics data (total searches, top queries, CTR, etc.).
    def get_analytics(project_id)
      @client.request(:get, "/api/v2/projects/#{url_encode(project_id)}/analytics")
    end

    # ── Usage ───────────────────────────────────────────────────────────────

    # Get usage data for the project.
    #
    # GET /api/v2/projects/{projectId}/usage
    #
    # @param project_id [String] The project ID.
    # @return [Hash] Usage data with rows of type/count/since.
    def get_usage(project_id)
      @client.request(:get, "/api/v2/projects/#{url_encode(project_id)}/usage")
    end

    private

    def url_encode(str)
      URI.encode_www_form_component(str.to_s)
    end
  end
end
